### PR TITLE
Add SIM115 ruff rule

### PR DESCRIPTION
**Pull request by TomeHirata**
[Open in GitHub Codespaces](https://codespaces.new/TomeHirata/mlflow/pull/9890?quickstart=1)

### Related Issues/PRs

https://github.com/mlflow/mlflow/issues/9834

### What changes are proposed in this pull request?

This PR is for adding SIM115 open-file-with-context-handler to the ruff config.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/recipes/steps/evaluate.py [hunk 1]
```diff
@@ -136,107 +136,110 @@ def my_warn(*args, **kwargs):
...
+                f.write(message)
 
         original_warn = warnings.warn
         warnings.warn = my_warn
         try:
             import pandas as pd
 
-            open(os.path.join(output_directory, "warning_logs.txt"), "w")
+            with open(os.path.join(output_directory, "warning_logs.txt"), "w"):
+                self._validate_validation_criteria()
```
Comment (harupy) [inline mlflow/recipes/steps/evaluate.py:148]:
```suggestion
            with open(os.path.join(output_directory, "warning_logs.txt"), "w"):
                pass

            self._validate_validation_criteria()
```

File: mlflow/utils/file_utils.py [hunk 2]
```diff
@@ -753,7 +753,7 @@ def run_download(range_start, range_end):
     # Create file if it doesn't exist or erase the contents if it does. We should do this here
     # before sending to the workers so they can each individually seek to their respective positions
     # and write chunks without overwriting.
-    open(download_path, "w").close()
+    open(download_path, "w").close()  # noqa: SIM115
```
Comment (harupy) [inline mlflow/utils/file_utils.py:756]:
```suggestion
    with open(download_path, "w"):
        pass
```

File: tests/utils/test_file_utils.py [hunk 3]
```diff
@@ -213,7 +213,7 @@ def test_mkdir(tmp_path):
 
     # raises if it exists already but is a file
     dummy_file_path = str(tmp_path.joinpath("dummy_file"))
-    open(dummy_file_path, "a").close()
+    open(dummy_file_path, "a").close()  # noqa: SIM115
```
Comment (harupy) [inline tests/utils/test_file_utils.py:216]:
```suggestion
    with open(dummy_file_path, "a"):
        pass
```

File: mlflow/recipes/artifacts.py [hunk 4]
```diff
@@ -154,7 +154,8 @@ def path(self):
 
     def load(self):
         if os.path.exists(self._path):
-            return open(self._path).read()
+            with open(self._path) as f:
+                return f.read()
```
Comment (harupy) [inline mlflow/recipes/artifacts.py:158]: /ai Is `open(...).read()` unsafe?

File: tests/recipes/test_train_step.py [hunk 5]
```diff
@@ -479,7 +479,8 @@ def test_train_step_with_tuning_best_parameters(
     assert "penalty" in best_params_yaml
     assert "eta0" in best_params_yaml
 
-    run_id = open(train_step_output_dir / "run_id").read()
+    with open(train_step_output_dir / "run_id") as f:
+        run_id = f.read()
```
Comment (harupy) [inline tests/recipes/test_train_step.py:483]:
```suggestion
    run_id = train_step_output_dir.joinpath("run_id").read_text()
```

looks like `train_step_output_dir` is a `Path` object. We can use `read_text`.

File: mlflow/recipes/steps/evaluate.py (hunk 1 above)
Comment (TomeHirata) [reply mlflow/recipes/steps/evaluate.py:148]: Oh, I thought we wanted to keep opening the `warning_logs.txt`. Is it just for creating a log file? Then, let me change here.

**Comment by TomeHirata** [comment]
@harupy Thank you for the suggestions. Could you please take another look when you have time?

File: mlflow/recipes/steps/train.py [hunk 6]
```diff
@@ -291,330 +292,334 @@ def my_warn(*args, **kwargs):
 
             from mlflow.models import infer_signature
 
-            open(os.path.join(output_directory, "warning_logs.txt"), "w")
+            with open(os.path.join(output_directory, "warning_logs.txt"), "w"):
+                apply_recipe_tracking_config(self.tracking_config)
```
Comment (harupy) [inline mlflow/recipes/steps/train.py:296]:
```suggestion
            with open(os.path.join(output_directory, "warning_logs.txt"), "w"):
                pass
             
            apply_recipe_tracking_config(self.tracking_config)
```

**Comment by harupy** [comment]
@CodiumAI-Agent /review

File: mlflow/recipes/steps/train.py (hunk 6 above)
Comment (TomeHirata) [reply mlflow/recipes/steps/train.py]: Thank you, updated here as well.

File: mlflow/recipes/steps/evaluate.py [hunk 7]
```diff
@@ -224,7 +226,8 @@ def my_warn(*args, **kwargs):
                     eval_result.save(os.path.join(output_directory, f"eval_{dataset_name}"))
                     eval_metrics[dataset_name] = {
                         transform_multiclass_metric(
-                            strip_prefix(k, evaluator_config["metric_prefix"]), self.extended_task
+                            strip_prefix(k, evaluator_config["metric_prefix"]),
+                            self.extended_task,
```
Comment (harupy) [inline mlflow/recipes/steps/evaluate.py:230]:
```suggestion
                            strip_prefix(k, evaluator_config["metric_prefix"]), self.extended_task
```

File: mlflow/recipes/steps/train.py [hunk 8]
```diff
@@ -323,7 +325,10 @@ def my_warn(*args, **kwargs):
                             f"skipping rebalancing."
                         )
 
-            X_train, y_train = train_df.drop(columns=[self.target_col]), train_df[self.target_col]
+            X_train, y_train = (
+                train_df.drop(columns=[self.target_col]),
+                train_df[self.target_col],
+            )
```
Comment (harupy) [inline mlflow/recipes/steps/train.py:331]:
```suggestion
            X_train, y_train = train_df.drop(columns=[self.target_col]), train_df[self.target_col]
```

File: mlflow/recipes/steps/train.py [hunk 9]
```diff
@@ -416,8 +421,8 @@ def my_warn(*args, **kwargs):
                     # Saving a temp model so that the output schema (signature) of the model's
                     # pyfunc representation can be inferred and included when logging the model
                     # to MLflow Tracking. Unfortunately, there is currently no easy way to infer
-                    # the model's signature without first saving a copy of it, and there is no easy
-                    # way to add an inferred signature to an existing model
+                    # the model's signature without first saving a copy of it, and there is
+                    # no easy way to add an inferred signature to an existing model
```
Comment (harupy) [inline mlflow/recipes/steps/train.py:425]:
```suggestion
                    # the model's signature without first saving a copy of it, and there is no easy
                    # way to add an inferred signature to an existing model
```

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**